### PR TITLE
Add thread initializer (from Python 3.7)

### PR DESCRIPTION
**Pull request by fahhem**

### CODE REVIEW

File: concurrent/futures/thread.py [hunk 1]
```diff
@@ -67,7 +67,9 @@ def run(self):
         else:
             self.future.set_result(result)
 
-def _worker(executor_reference, work_queue):
+def _worker(executor_reference, work_queue, initializer, initargs):
+    if initializer is not None:
+        initializer(*initargs)
```
Comment (agronholm) [inline concurrent/futures/thread.py]: This looks quite different from upstream code. For one, exceptions are not handled.

Comment (fahhem) [reply concurrent/futures/thread.py]: Ah, I wasn't aware this was a straight copy of upstream. I'll do that now

Comment (agronholm) [reply concurrent/futures/thread.py]: It was originally copied from upstream obviously but modified for Python 2. I try to stick as close to upstream as reasonably possible.

Comment (fahhem) [reply concurrent/futures/thread.py]: Okay, then I've copied the Python3 version of this, and pulled in the Broken* exceptions too

Comment (agronholm) [reply concurrent/futures/thread.py]: Good. If there are any tests upstream that cover the failure of the initializer, I'd like those added too.

**Comment by fahhem** [comment]
Hmm, the tests are significantly different from upstream (upstream seems to have improved them significantly).

Should I just re-backport the latest from 3.7 or just pull in thread initializers?

**Comment by agronholm** [comment]
> Hmm, the tests are significantly different from upstream (upstream seems to have improved them significantly).
> 
> Should I just re-backport the latest from 3.7 or just pull in thread initializers?

I'm not sure that the tests are fully applicable to the backport (as the upstream code is not). The initializer related test should be sufficient (as it seems you've already done).

**Comment by agronholm** [comment]
I'm going to have to fix the test run for py2.6 and make sure it passes before merging this.

**Comment by agronholm** [comment]
Thanks!

**Comment by agronholm** [comment]
Hm, now we have a partial implementation of [bpo-21423](https://bugs.python.org/issue21423). I don't suppose I could ask you to make a PR covering process pools too? I'm not comfortable making a release with a halfway implemented feature.

**Comment by fahhem** [comment]
Here's the PR: https://github.com/agronholm/pythonfutures/pull/94

**Pull request merged**